dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java

```diff
@@ -637,60 +568,48 @@ private String getFromSubQuery( TrackedEntityInstanceQueryParams params, boolean
-     * based on the params we are ordering by.
-     *
-     * @param params
-     * @return a string containing the columns to be used in the subquery SQL
-     *         projection, or empty string if no attributes nor order params are
-     *         used for ordering.
-     */
-    private String getSubqueryFromClauseOrderFields( TrackedEntityInstanceQueryParams params, String query )
-    {
-        StringBuilder orderAttributes = new StringBuilder();
+        LinkedHashSet<String> columns = new LinkedHashSet<>(
+            List.of( "TEI.trackedentityinstanceid", "TEI.uid", "TEI.created", "TEI.lastupdated",
+                "TEI.inactive", "TEI.trackedentitytypeid", "TEI.potentialduplicate", "TEI.deleted", "OU.uid as ou",
+                "OU.name as ouname " ) );
 
         for ( OrderParam orderParam : params.getOrders() )
         {
-            if ( isStaticColumn( orderParam.getField() ) )
+            Optional<TrackedEntityInstanceQueryParams.OrderColumn> orderColumn = getColumn( orderParam.getField() );
+
+            if ( orderColumn.isPresent() )
             {
-                orderAttributes.append( addStaticColumn( orderParam.getField(), query ) );
+                columns.add( orderColumn.get().getTableAlias() + "." + orderColumn.get().getColumn() );
             }
             else
             {
-                Optional<QueryItem> orderAttribute = params.getAttributesAndFilters().stream()
-                    .filter( queryItem -> orderParam.getField().equalsIgnoreCase( queryItem.getItemId() ) )
-                    .findFirst();
-
-                orderAttribute.ifPresent( queryItem -> orderAttributes
-                    .append( ", " )
-                    .append( statementBuilder.columnQuote( queryItem.getItemId() ) )
-                    .append( ".value AS " )
-                    .append( statementBuilder.columnQuote( queryItem.getItemId() ) ) );
+                if ( sortableAttributesAndFilters( params ).stream()
+                    .anyMatch( i -> i.getItem().getUid().equals( orderParam.getField() ) ) )
+                {
+                    columns.add( statementBuilder.columnQuote( orderParam.getField() ) + ".value AS " +
+                        statementBuilder.columnQuote( orderParam.getField() ) );
+                }
             }
         }
 
-        return orderAttributes.toString();
+        return "SELECT " + Strings.join( ", ", columns );
+    }
+
+    /**
+     * Get a list of QueryItem that contains sortable attributes also defined as
+     * filers
+     *
+     * @param params
+     * @return List of QueryItem
+     */
+    private List<QueryItem> sortableAttributesAndFilters( TrackedEntityInstanceQueryParams params )
+    {
+        List<String> ordersIdentifier = params.getOrders().stream()
+            .map( OrderParam::getField )
+            .collect( Collectors.toList() );
+        return params.getAttributesAndFilters().stream()
```

Comment:
same as above, why don't we avoid the creation of the list and map and filter in the same statement instead?